File: Nodejs/Railway_version/test/controllers/users_controller_test.js
require('../test_helper.js').controller('users', module.exports);

var sinon  = require('sinon');

function ValidAttributes () {
    return {
        username: 'unit_test',
        password: 'toto',
        email: 'unit_test@agbuilder.eu',
        home: '/users/unit_test/'
    };
}

function InvalidAttributes () {
    return {
        username: '',
        password: '',
        email: '',
        home: ''
    };
}

exports['users controller'] = {

    'GET new': function (test) {
        test.get('/users/new', function () {
            test.success();
            test.render('new');
            test.render('form.' + app.set('view engine'));
            test.done();
        });
    },

    'GET index': function (test) {
        test.get('/users', function () {
            test.success();
            test.render('index');
            test.done();
        });
    },

    'GET edit': function (test) {
        var find = User.find;
        User.find = sinon.spy(function (id, callback) {
            callback(null, new User);
        });
        test.get('/users/42/edit', function () {
            test.ok(User.find.calledWith('42'));
            User.find = find;
            test.success();
            test.render('edit');
            test.done();
        });
    },

    'GET show': function (test) {
        var find = User.find;
        User.find = sinon.spy(function (id, callback) {
            callback(null, new User);
        });
        test.get('/users/42', function (req, res) {
            test.ok(User.find.calledWith('42'));
            User.find = find;
            test.success();
            test.render('show');
            test.done();
        });
    },

    'POST create': function (test) {
        var newUser = new ValidAttributes;
        newUser.password_confirm = newUser.password;

        sinon.spy(User, 'create');
        test.post('/users', {User: newUser}, function () {
            test.redirect('/users');
            test.flash('info');
            test.strictEqual(User.create.calledOnce, true);
            User.create.restore();
            test.done();
        });
    },

    'POST create fail': function (test) {
        var newUser = new ValidAttributes;
        newUser.password_confirm = "wrong password";

        sinon.spy(User, 'create');
        test.post('/users', {User: newUser}, function () {
            test.redirect('/users/new');
            test.flash('error');
            test.strictEqual(User.create.called, false);
            User.create.restore();
            test.done();
        });
    },

    'PUT update': function (test) {
        test.put('/users/1', {User: new ValidAttributes}, function () {
            test.redirect('/users/1');
            test.flash('info');
            test.done();
        });
    },

    'PUT update fail': function (test) {
        test.put('/users/1', {User: new InvalidAttributes}, function () {
            test.success();
            test.render('edit');
            test.flash('error');
            test.done();
        });
    },

    'DELETE destroy': function (test) {
        test.del('/users/1', function() {
            test.success();
            test.flash('info');
            test.done();
        });
    },

    'DELETE destroy fail': function (test) {
        test.del('/users/42', function() {
            test.redirect('/users');
            test.flash('error');
            test.done();
        });
    }
};

File: Rails/AGB/test/functional/users_controller_test.rb
require 'test_helper'

class UsersControllerTest < ActionController::TestCase
  setup do
    @user = users(:one)
    @update = {
      name: "Moreau",
      password: "toto42",
      email: "moreau@agb.com"
    }
  end

  test "should create user" do
    @update.password_digest = BCrypt::Password.create(@update.password)
    assert_difference('User.count') do
      post :create, user: @update
    end

    assert_redirected_to user_path(assigns(:user))
  end

  # test "should show user" do
  #   get :show, id: @user
  #   assert_response :success
  # end

  # test "should get edit" do
  #   get :edit, id: @user
  #   assert_response :success
  # end

  # test "should update user" do
  #   put :update, id: @user, user: { email: @user.email, name: @user.name, password: @user.password }
  #   assert_redirected_to user_path(assigns(:user))
  # end

  # test "should destroy user" do
  #   assert_difference('User.count', -1) do
  #     delete :destroy, id: @user
  #   end

  #   assert_redirected_to users_path
  # end
end
